Some improvements to optimizer documentation (#11360)
ethereum/solidity: docs/internals/optimizer.rst

*************

The Solidity compiler uses two different optimizer modules: The "old" optimizer
that operates at the opcode level and the "new" optimizer that operates on Yul IR code.

The opcode-based optimizer applies a set of `simplification rules <https://github.com/ethereum/solidity/blob/develop/libevmasm/RuleList.h>`_
to opcodes. It also combines equal code sets and removes unused code.

The Yul-based optimizer is much more powerful, because it can work across function

> calls. For example,  arbitrary jumps are not possible in Yul, so it is

Reviewer: ```suggestion
calls. For example, arbitrary jumps are not possible in Yul, so it is
```

Differences between Optimized and Non-Optimized Code
====================================================

Generally, the most visible difference is that constant expressions are evaluated at compile time.

> When it comes to the ASM output, one can also notice a reduction of redundant

Reviewer: Redundant might be too strong here. I think this line is referring to Block Deduplicator. This optimizer step would look for code blocks that are the same, removes all except one and replace jumps to blocks that were removed to the one block that remains.

> When it comes to the ASM output, one can also notice a reduction of redundant

Reviewer: ```suggestion
When it comes to the ASM output, one can also notice a reduction of equivalent or duplicate
```

The number of runs (``--optimize-runs``) specifies roughly how often each opcode of the
deployed code will be executed across the life-time of the contract. This means it is a
trade-off parameter between code size (deploy cost) and code execution cost (cost after deployment).

> A "runs" parameter of "1" will produce short but expensive code. The maximum value of ``2**32-1`` will
> produce longer but more gas efficient code.

Reviewer: I feel that currently it looks as if only two values are supported.

```suggestion
A "runs" parameter of "1" will produce short but expensive code. In contrast, a larger "runs" parameter will produce longer by more gas efficient code. The maximum value of the parameter is ``2**32-1``.
```

Opcode-Based Optimizer Module
=============================

The opcode-based optimizer module operates on assembly code. It splits the
sequence of instructions into basic blocks at ``JUMPs`` and ``JUMPDESTs``.
Inside these blocks, the optimizer analyzes the instructions and records every modification to the stack,
memory, or storage as an expression which consists of an instruction and
a list of arguments which are pointers to other expressions. 

Additionally, the opcode-based optimizer
uses a component called "CommonSubexpressionEliminator" that, amongst other
tasks, finds expressions that are always equal (on every input) and combines
them into an expression class. It first tries to find each new
expression in a list of already known expressions. If no such matches are found,
it simplifies the expression according to rules like
``constant + constant = sum_of_constants`` or ``X * 1 = X``. Since this is
a recursive process, we can also apply the latter rule if the second factor
is a more complex expression which we know always evaluates to one.

> Modifications to storage and memory locations must erase knowledge about

Reviewer: ```suggestion
Some optimizer steps symbolically track memory and storage. Therefore,
modifications to storage and memory locations must erase knowledge about
```

Additionally, the opcode-based optimizer
uses a component called "CommonSubexpressionEliminator" that, amongst other
tasks, finds expressions that are always equal (on every input) and combines
them into an expression class. It first tries to find each new
expression in a list of already known expressions. If no such matches are found,
it simplifies the expression according to rules like
``constant + constant = sum_of_constants`` or ``X * 1 = X``. Since this is
a recursive process, we can also apply the latter rule if the second factor
is a more complex expression which we know always evaluates to one.

Certain optimizer steps symbolically tracks the storage and memory locations. Therefore
modifications to storage and memory locations, of say location ``l``, must erase knowledge about
storage or memory locations which are not known to be different from ``l``. For example, If we first
write to a symbolic storage location ``x`` and then to the symbolic storage location ``y``. In
general, ``y`` could overwrite ``x`` if ``y`` and ``x`` are the same storage slot. However, if the
optimizer can infer that the value of the expression ``x - y`` is non-zero, this means that the

> knowledge about storage at the symbolic location ``x`` can be kept.

Reviewer: Updated this. @leonardoalt Perhaps a review of at least this part?

Reviewer 2: Hmm I don't understand how it arrives at the implication that knowledge can be kept if the expression results in non-zero...

Opcode-Based Optimizer Module
=============================

> The opcode-based optimizer module operates on assembly code. It splits the

Author: This paragraph was a big block. I broke it up into what seemed to me like reasonable sub paragraphs.

sequence of instructions into basic blocks at ``JUMPs`` and ``JUMPDESTs``.
Inside these blocks, the optimizer analyzes the instructions and records every modification to the stack,
memory, or storage as an expression which consists of an instruction and
a list of arguments which are pointers to other expressions. 

Additionally, the opcode-based optimizer
uses a component called "CommonSubexpressionEliminator" that, amongst other
tasks, finds expressions that are always equal (on every input) and combines
them into an expression class. It first tries to find each new
expression in a list of already known expressions. If no such matches are found,
it simplifies the expression according to rules like
``constant + constant = sum_of_constants`` or ``X * 1 = X``. Since this is
a recursive process, we can also apply the latter rule if the second factor
is a more complex expression which we know always evaluates to one.

> Modifications to storage and memory locations must erase knowledge about

Author: I really struggled to understand this paragraph. I think it's saying something about a function that takes arguments x, and y, and does `SSTORE(X, VAL1)`, and  `SSTORE(Y, VAL2)`, but then I'm not sure what the optimizer does about that.

Reviewer: The optimizer tries to symbolically track memory and storage. As an application, this is how Keccak-256 may be evaluated during compile time. For example, consider the sequence:

```
 [PUSH 32, PUSH 0, CALLDATALOAD, PUSH 100,  DUP2, MSTORE, KECCAK256]
```

or the equivalent Yul
```
let x := calldataload(0)
mstore(x, 100)
let value := keccak256(x, 32)
```

In this case, the optimizer has to 'symbolically' track the value at a memory location `calldataload(0)` and then realize that the Keccak-256 has can be evaluated at compile time. This only works if there is no other instruction that modifies memory in between the `mstore` and `keccak256`. So if there is an instruction that writes to memory (or storage), then we need to erase the knowledge of the current memory (or storage). There is, however, an exception to this erasing, where we can easily see that the instruction doesn't write to a certain location.

For example,

```
let x := calldataload(0)
mstore(x, 100)
/// Current knowledge memory location x -> 100
let y := add(x, 32)
/// Does not clear the knowledge that x -> 100, since y does not write to [x, x + 32)
mstore(y, 200)
```

-----

I admit that the current paragraph is easy to understand if you already know how it works and rather confusing otherwise. Do you have a suggestion on rewording it?

Opcode-Based Optimizer Module
=============================

The opcode-based optimizer module operates on assembly code. It splits the
sequence of instructions into basic blocks at ``JUMPs`` and ``JUMPDESTs``.
Inside these blocks, the optimizer analyzes the instructions and records every modification to the stack,
memory, or storage as an expression which consists of an instruction and
a list of arguments which are pointers to other expressions.

Additionally, the opcode-based optimizer
uses a component called "CommonSubexpressionEliminator" that, amongst other
tasks, finds expressions that are always equal (on every input) and combines
them into an expression class. It first tries to find each new
expression in a list of already known expressions. If no such matches are found,
it simplifies the expression according to rules like
``constant + constant = sum_of_constants`` or ``X * 1 = X``. Since this is
a recursive process, we can also apply the latter rule if the second factor
is a more complex expression which we know always evaluates to one.

> Certain optimizer steps symbolically tracks the storage and memory locations. For example, this

Reviewer 3: ```suggestion
Certain optimizer steps symbolically track the storage and memory locations. For example, this
```

information is used to compute Keccak-256 hashes that can be evaluated during compile time. Consider
the sequence:

::

  PUSH 32
  PUSH 0
  CALLDATALOAD
  PUSH 100
  DUP2
  MSTORE
  KECCAK256

or the equivalent Yul

::

  let x := calldataload(0)
  mstore(x, 100)
  let value := keccak256(x, 32)

> In this case, the optimizer has to 'symbolically' track the value at a memory location

Reviewer 3: ```suggestion
In this case, the optimizer has to symbolically track the value at a memory location
```
(same sentence but no quotes above)

> ``calldataload(0)`` and then realize that the Keccak-256 hash can be evaluated at compile time. This

Reviewer 3: Hmm these last 2 sentences basically repeat the sentence before the example code.

> only works if there is no other instruction that modifies memory in between the ``mstore`` and

Reviewer 3: ```suggestion
only works if there is no other instruction that modifies memory between the ``mstore`` and
```

``keccak256``. So if there is an instruction that writes to memory (or storage), then we need to
erase the knowledge of the current memory (or storage). There is, however, an exception to this
erasing, when we can easily see that the instruction doesn't write to a certain location.

For example,

::

  let x := calldataload(0)
  mstore(x, 100)
  // Current knowledge memory location x -> 100
  let y := add(x, 32)
  // Does not clear the knowledge that x -> 100, since y does not write to [x, x + 32)
  mstore(y, 200)
  // This Keccak-256 can now be evaluated
  let value := keccak256(x, 32)

Therefore, modifications to storage and memory locations, of say location ``l``, must erase
knowledge about storage or memory locations which may be equal to ``l``. More specifically, for
storage, the optimizer has to erase all knowledge of symbolic locations, that may be equal to ``l``
and for memory, the optimizer has to erase all knowledge of symbolic locations that may not be at

> least 32 bytes away. This is done by computing the value ``sub(l, m)``. For storage, if this value

Reviewer 3: What's `m`?

a list of arguments which are pointers to other expressions.

Additionally, the opcode-based optimizer
uses a component called "CommonSubexpressionEliminator" that, amongst other
tasks, finds expressions that are always equal (on every input) and combines
them into an expression class. It first tries to find each new
expression in a list of already known expressions. If no such matches are found,
it simplifies the expression according to rules like
``constant + constant = sum_of_constants`` or ``X * 1 = X``. Since this is
a recursive process, we can also apply the latter rule if the second factor
is a more complex expression which we know always evaluates to one.

Certain optimizer steps symbolically track the storage and memory locations. For example, this
information is used to compute Keccak-256 hashes that can be evaluated during compile time. Consider
the sequence:

::

  PUSH 32
  PUSH 0
  CALLDATALOAD
  PUSH 100
  DUP2
  MSTORE
  KECCAK256

or the equivalent Yul

::

  let x := calldataload(0)
  mstore(x, 100)
  let value := keccak256(x, 32)

In this case, the optimizer tracks the value at a memory location ``calldataload(0)`` and then

> realize that the Keccak-256 hash can be evaluated at compile time. This only works if there is no

Reviewer 3: ```suggestion
realizes that the Keccak-256 hash can be evaluated at compile time. This only works if there is no
```

Reviewer: Done!
